From 3da8d6b12d48b5a4401fbf5d1e75e867928dc835 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Sat, 30 May 2026 09:58:25 +0200
Subject: [PATCH 05/22] oplus: Introduce bare minimum hbp setup tool for
 recovery touch support

Change-Id: Ibc29b4d764643caad71f43511f887e86619bde74
(cherry picked from commit bdec56c751e82302d1de4760048248d7cf055a43)

Ledger: hardware_oplus.md Accepted Group A row 6 (bdec56c7)
Change-Id: I0e0b18079f0008c5af279b766db78cb9209708aa
(cherry picked from commit e12fed2319acea5b4405a6c71fb97828ff646477)
(cherry picked from commit 560e91b900adb641ee6a5de9b9087f189be93b4c)
---
 hbp-daemon/Android.bp                    | 16 +++++
 hbp-daemon/hbp-setup.rc                  |  6 ++
 hbp-daemon/main.cpp                      | 23 +++++++
 kernel-headers/include/oplus/oplus_hbp.h | 78 ++++++++++++++++++++++++
 sepolicy/qti/vendor/hbp_setup.te         |  7 +++
 5 files changed, 130 insertions(+)
 create mode 100644 hbp-daemon/Android.bp
 create mode 100644 hbp-daemon/hbp-setup.rc
 create mode 100644 hbp-daemon/main.cpp
 create mode 100644 kernel-headers/include/oplus/oplus_hbp.h
 create mode 100644 sepolicy/qti/vendor/hbp_setup.te

diff --git a/hbp-daemon/Android.bp b/hbp-daemon/Android.bp
new file mode 100644
index 00000000..3233702d
--- /dev/null
+++ b/hbp-daemon/Android.bp
@@ -0,0 +1,16 @@
+//
+// SPDX-FileCopyrightText: 2026 The LineageOS Project
+// SPDX-License-Identifier: Apache-2.0
+//
+
+cc_binary {
+    name: "hbp-setup",
+    init_rc: ["hbp-setup.rc"],
+    recovery: true,
+    srcs: [
+        "main.cpp",
+    ],
+    header_libs: [
+        "kernel_headers.oplus",
+    ],
+}
diff --git a/hbp-daemon/hbp-setup.rc b/hbp-daemon/hbp-setup.rc
new file mode 100644
index 00000000..1e605f55
--- /dev/null
+++ b/hbp-daemon/hbp-setup.rc
@@ -0,0 +1,6 @@
+service vendor.hbp-setup /system/bin/hbp-setup
+    class hal
+    user root
+    group root
+    seclabel u:r:hbp_setup:s0
+    oneshot
diff --git a/hbp-daemon/main.cpp b/hbp-daemon/main.cpp
new file mode 100644
index 00000000..f684671f
--- /dev/null
+++ b/hbp-daemon/main.cpp
@@ -0,0 +1,23 @@
+/*
+ * SPDX-FileCopyrightText: 2026 The LineageOS Project
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#undef NDEBUG
+
+#include <assert.h>
+#include <fcntl.h>
+#include <oplus/oplus_hbp.h>
+
+int main() {
+    auto fd = open("/dev/hbp0", O_RDWR);
+    assert(fd != -1);
+
+    usr_data data{};
+    data.val = 1;
+
+    auto ret = ioctl(fd, HBP_IOCTRL_IRQ_ENABLE, &data);
+    assert(ret == 0);
+
+    return 0;
+}
diff --git a/kernel-headers/include/oplus/oplus_hbp.h b/kernel-headers/include/oplus/oplus_hbp.h
new file mode 100644
index 00000000..f901933f
--- /dev/null
+++ b/kernel-headers/include/oplus/oplus_hbp.h
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2026 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/ioctl.h>
+
+#define HBP_IOCTRL_GROUP 0xC5
+
+#define HBP_IOCTRL_IRQ_ENABLE _IO(HBP_IOCTRL_GROUP, 0x0A)
+
+#define MAX_POWER_SEQ (10)
+
+enum power_type { POWER_AVDD = 0x1000, POWER_VDDI, POWER_RESET, POWER_BUS, POWER_MAX };
+
+struct power_sequeue {
+    enum power_type type;
+    bool en;
+    uint32_t msleep;
+};
+
+union usr_data {
+    int64_t val;
+
+    struct {
+        void* tx;
+        void* rx;
+        int32_t len;
+    } rw;
+
+    struct {
+        void* data;
+        size_t size;
+    } frame;
+
+    struct {
+        uint8_t state;
+        int x;
+        int y;
+        int touch_early_down_flag;
+        long is_touch_fp_area_Cnt;
+        int tp_firmware_time;
+    } ifp;
+
+    struct power_sequeue sq[MAX_POWER_SEQ];
+
+    struct {
+        uint8_t mode;
+        uint8_t bits_per_word;
+        int speed;
+    } spi_setup;
+
+    struct {
+        bool filmed;
+        int level;
+        bool trusty;
+    } film;
+
+    struct {
+        void* info;
+        size_t info_size;
+    } health_info;
+};
diff --git a/sepolicy/qti/vendor/hbp_setup.te b/sepolicy/qti/vendor/hbp_setup.te
new file mode 100644
index 00000000..3f67f362
--- /dev/null
+++ b/sepolicy/qti/vendor/hbp_setup.te
@@ -0,0 +1,7 @@
+recovery_only(`
+type hbp_setup, domain, coredomain;
+
+domain_trans(init, rootfs, hbp_setup)
+
+allow hbp_setup oplus_touchdaemon_device:chr_file rw_file_perms;
+')
